File: interactive-messages/csharp/InteractiveMessages.cs
/**
 * SendSeven API - Interactive Messages Example
 *
 * Demonstrates how to send interactive messages (buttons, lists, quick replies)
 * using the SendSeven API.
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace SendSeven.Examples
{
    class InteractiveMessages
    {
        // Configuration from environment
        private static string? ApiToken;
        private static string? TenantId;
        private static string? ApiUrl;
        private static string? ChannelId;
        private static string? ContactId;

        private static readonly HttpClient httpClient = new HttpClient();
        private static readonly JsonSerializerOptions jsonOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
            WriteIndented = false
        };

        static async Task Main(string[] args)
        {
            LoadEnv();

            // Validate configuration
            if (string.IsNullOrEmpty(ApiToken))
            {
                Console.WriteLine("Error: SENDSEVEN_API_TOKEN environment variable is required");
                Environment.Exit(1);
            }

            if (string.IsNullOrEmpty(TenantId))
            {
                Console.WriteLine("Error: SENDSEVEN_TENANT_ID environment variable is required");
                Environment.Exit(1);
            }

            if (string.IsNullOrEmpty(ChannelId))
            {
                Console.WriteLine("Error: CHANNEL_ID environment variable is required");
                Environment.Exit(1);
            }

            if (string.IsNullOrEmpty(ContactId))
            {
                Console.WriteLine("Error: CONTACT_ID environment variable is required");
                Environment.Exit(1);
            }

            // 1. Check channel capabilities first
            Console.WriteLine($"Checking capabilities for channel: {ChannelId}");
            try
            {
                var capabilities = await CheckChannelCapabilities(ChannelId);
                Console.WriteLine($"Channel type: {GetJsonValue(capabilities, "channel_type", "unknown")}");

                if (capabilities.TryGetProperty("capabilities", out var caps))
                {
                    Console.WriteLine($"  Buttons: {GetJsonValue(caps, "interactive_buttons", "false")}");
                    Console.WriteLine($"  Lists: {GetJsonValue(caps, "interactive_lists", "false")}");
                    Console.WriteLine($"  Quick Replies: {GetJsonValue(caps, "quick_replies", "false")}");
                }
                Console.WriteLine();
            }
            catch (Exception e)
            {
                Console.WriteLine($"Warning: Could not check capabilities: {e.Message}");
                Console.WriteLine("Proceeding anyway...");
                Console.WriteLine();
            }

            // 2. Send a button message
            Console.WriteLine("Sending button message...");
            try
            {
                var buttons = new List<Dictionary<string, string>>
                {
                    new() { { "id", "yes" }, { "title", "Yes" } },
                    new() { { "id", "no" }, { "title", "No" } },
                    new() { { "id", "maybe" }, { "title", "Maybe Later" } }
                };

                var message = await SendButtonMessage(
                    ChannelId,
                    ContactId,
                    "Would you like to proceed with your order?",
                    buttons
                );

                Console.WriteLine("Button message sent successfully!");
                Console.WriteLine($"  ID: {GetJsonValue(message, "id")}");
                Console.WriteLine($"  Status: {GetJsonValue(message, "status")}");
                Console.WriteLine();
            }
            catch (Exception e)
            {
                Console.WriteLine($"Button message failed: {e.Message}");
                Console.WriteLine();
            }

            // 3. Send a list message
            Console.WriteLine("Sending list message...");
            try
            {
                var sections = new List<Dictionary<string, object>>
                {
                    new()
                    {
                        { "title", "Electronics" },
                        { "rows", new List<Dictionary<string, string>>
                            {
                                new() { { "id", "phones" }, { "title", "Phones" }, { "description", "Latest smartphones" } },
                                new() { { "id", "laptops" }, { "title", "Laptops" }, { "description", "Portable computers" } }
                            }
                        }
                    },
                    new()
                    {
                        { "title", "Accessories" },
                        { "rows", new List<Dictionary<string, string>>
                            {
                                new() { { "id", "cases" }, { "title", "Cases" }, { "description", "Protective cases" } },
                                new() { { "id", "chargers" }, { "title", "Chargers" }, { "description", "Fast chargers" } }
                            }
                        }
                    }
                };

                var message = await SendListMessage(
                    ChannelId,
                    ContactId,
                    "Browse our product catalog:",
                    "View Products",
                    sections
                );

                Console.WriteLine("List message sent successfully!");
                Console.WriteLine($"  ID: {GetJsonValue(message, "id")}");
                Console.WriteLine($"  Status: {GetJsonValue(message, "status")}");
                Console.WriteLine();
            }
            catch (Exception e)
            {
                Console.WriteLine($"List message failed: {e.Message}");
                Console.WriteLine();
            }

            // 4. Send a quick reply message
            Console.WriteLine("Sending quick reply message...");
            try
            {
                var quickReplies = new List<Dictionary<string, string>>
                {
                    new() { { "id", "excellent" }, { "title", "Excellent" } },
                    new() { { "id", "good" }, { "title", "Good" } },
                    new() { { "id", "poor" }, { "title", "Poor" } }
                };

                var message = await SendQuickReplyMessage(
                    ChannelId,
                    ContactId,
                    "How would you rate our service today?",
                    quickReplies
                );

                Console.WriteLine("Quick reply message sent successfully!");
                Console.WriteLine($"  ID: {GetJsonValue(message, "id")}");
                Console.WriteLine($"  Status: {GetJsonValue(message, "status")}");
            }
            catch (Exception e)
            {
                Console.WriteLine($"Quick reply message failed: {e.Message}");
            }
        }

        /// <summary>
        /// Load environment variables from .env file
        /// </summary>
        private static void LoadEnv()
        {
            // Try to load from .env file
            var envPath = Path.Combine(Directory.GetCurrentDirectory(), ".env");
            if (File.Exists(envPath))
            {
                foreach (var line in File.ReadAllLines(envPath))
                {
                    var trimmedLine = line.Trim();
                    if (string.IsNullOrEmpty(trimmedLine) || trimmedLine.StartsWith("#"))
                        continue;

                    var idx = trimmedLine.IndexOf('=');
                    if (idx > 0)
                    {
                        var key = trimmedLine[..idx].Trim();
                        var value = trimmedLine[(idx + 1)..].Trim();

                        // Remove quotes if present
                        if ((value.StartsWith("\"") && value.EndsWith("\"")) ||
                            (value.StartsWith("'") && value.EndsWith("'")))
                        {
                            value = value[1..^1];
                        }

                        Environment.SetEnvironmentVariable(key, value);
                    }
                }
            }

            // Load configuration
            ApiToken = Environment.GetEnvironmentVariable("SENDSEVEN_API_TOKEN");
            TenantId = Environment.GetEnvironmentVariable("SENDSEVEN_TENANT_ID");
            ApiUrl = Environment.GetEnvironmentVariable("SENDSEVEN_API_URL") ?? "https://api.sendseven.com/api/v1";
            ChannelId = Environment.GetEnvironmentVariable("CHANNEL_ID");
            ContactId = Environment.GetEnvironmentVariable("CONTACT_ID");
        }

        /// <summary>
        /// Get a value from a JsonElement
        /// </summary>
        private static string GetJsonValue(JsonElement element, string propertyName, string defaultValue = "")
        {
            if (element.TryGetProperty(propertyName, out var prop))
            {
                return prop.ValueKind switch
                {
                    JsonValueKind.String => prop.GetString() ?? defaultValue,
                    JsonValueKind.True => "true",
                    JsonValueKind.False => "false",
                    JsonValueKind.Number => prop.GetRawText(),
                    _ => prop.GetRawText()
                };
            }
            return defaultValue;
        }

        /// <summary>
        /// Make an HTTP request to the API
        /// </summary>
        private static async Task<JsonElement> MakeRequest(HttpMethod method, string url, object? body = null)
        {
            var request = new HttpRequestMessage(method, url);
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", ApiToken);
            request.Headers.Add("X-Tenant-ID", TenantId);

            if (body != null)
            {
                var json = JsonSerializer.Serialize(body);
                request.Content = new StringContent(json, Encoding.UTF8, "application/json");
            }

            var response = await httpClient.SendAsync(request);
            var responseBody = await response.Content.ReadAsStringAsync();

            if (!response.IsSuccessStatusCode)
            {
                throw new HttpRequestException($"HTTP {(int)response.StatusCode}: {responseBody}");
            }

            return JsonDocument.Parse(responseBody).RootElement;
        }

        /// <summary>
        /// Check what interactive message types a channel supports
        /// </summary>
        private static async Task<JsonElement> CheckChannelCapabilities(string channelId)
        {
            var url = $"{ApiUrl}/channels/{channelId}/capabilities";
            return await MakeRequest(HttpMethod.Get, url);
        }

        /// <summary>
        /// Send a button message to a contact
        /// </summary>
        private static async Task<JsonElement> SendButtonMessage(
            string channelId,
            string contactId,
            string body,
            List<Dictionary<string, string>> buttons)
        {
            var url = $"{ApiUrl}/messages/send/interactive";

            var payload = new Dictionary<string, object>
            {
                { "channel_id", channelId },
                { "contact_id", contactId },
                { "type", "buttons" },
                { "body", body },
                { "buttons", buttons }
            };

            return await MakeRequest(HttpMethod.Post, url, payload);
        }

        /// <summary>
        /// Send a list message with sections to a contact
        /// </summary>
        private static async Task<JsonElement> SendListMessage(
            string channelId,
            string contactId,
            string body,
            string buttonText,
            List<Dictionary<string, object>> sections)
        {
            var url = $"{ApiUrl}/messages/send/interactive";

            var payload = new Dictionary<string, object>
            {
                { "channel_id", channelId },
                { "contact_id", contactId },
                { "type", "list" },
                { "body", body },
                { "button_text", buttonText },
                { "sections", sections }
            };

            return await MakeRequest(HttpMethod.Post, url, payload);
        }

        /// <summary>
        /// Send a quick reply message to a contact
        /// </summary>
        private static async Task<JsonElement> SendQuickReplyMessage(
            string channelId,
            string contactId,
            string body,
            List<Dictionary<string, string>> buttons)
        {
            var url = $"{ApiUrl}/messages/send/interactive";

            var payload = new Dictionary<string, object>
            {
                { "channel_id", channelId },
                { "contact_id", contactId },
                { "type", "quick_reply" },
                { "body", body },
                { "buttons", buttons }
            };

            return await MakeRequest(HttpMethod.Post, url, payload);
        }
    }
}

File: interactive-messages/go/interactive_messages.go
// SendSeven API - Interactive Messages Example
//
// Demonstrates how to send interactive messages (buttons, lists, quick replies)
// using the SendSeven API.
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"

	"github.com/joho/godotenv"
)

// Configuration from environment
var (
	apiToken  string
	tenantID  string
	apiURL    string
	channelID string
	contactID string
)

// Button represents an interactive button
type Button struct {
	ID    string `json:"id"`
	Title string `json:"title"`
}

// ListRow represents a row in a list section
type ListRow struct {
	ID          string `json:"id"`
	Title       string `json:"title"`
	Description string `json:"description,omitempty"`
}

// ListSection represents a section in a list message
type ListSection struct {
	Title string    `json:"title"`
	Rows  []ListRow `json:"rows"`
}

// ChannelCapabilities represents the capabilities of a channel
type ChannelCapabilities struct {
	ChannelID    string                 `json:"channel_id"`
	ChannelType  string                 `json:"channel_type"`
	Capabilities map[string]interface{} `json:"capabilities"`
}

// Message represents a message response
type Message struct {
	ID     string `json:"id"`
	Status string `json:"status"`
}

// InteractiveRequest represents an interactive message request
type InteractiveRequest struct {
	ChannelID  string        `json:"channel_id"`
	ContactID  string        `json:"contact_id"`
	Type       string        `json:"type"`
	Body       string        `json:"body"`
	ButtonText string        `json:"button_text,omitempty"`
	Buttons    []Button      `json:"buttons,omitempty"`
	Sections   []ListSection `json:"sections,omitempty"`
}

func init() {
	// Load environment variables from .env file
	godotenv.Load()

	apiToken = os.Getenv("SENDSEVEN_API_TOKEN")
	tenantID = os.Getenv("SENDSEVEN_TENANT_ID")
	apiURL = os.Getenv("SENDSEVEN_API_URL")
	if apiURL == "" {
		apiURL = "https://api.sendseven.com/api/v1"
	}
	channelID = os.Getenv("CHANNEL_ID")
	contactID = os.Getenv("CONTACT_ID")
}

// makeRequest makes an HTTP request to the API
func makeRequest(method, url string, body interface{}) ([]byte, error) {
	var reqBody io.Reader
	if body != nil {
		jsonData, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal request body: %w", err)
		}
		reqBody = bytes.NewBuffer(jsonData)
	}

	req, err := http.NewRequest(method, url, reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+apiToken)
	req.Header.Set("X-Tenant-ID", tenantID)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(respBody))
	}

	return respBody, nil
}

// checkChannelCapabilities checks what interactive message types a channel supports
func checkChannelCapabilities(channelID string) (*ChannelCapabilities, error) {
	url := fmt.Sprintf("%s/channels/%s/capabilities", apiURL, channelID)

	respBody, err := makeRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	var capabilities ChannelCapabilities
	if err := json.Unmarshal(respBody, &capabilities); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &capabilities, nil
}

// sendButtonMessage sends a button message to a contact
func sendButtonMessage(channelID, contactID, body string, buttons []Button) (*Message, error) {
	url := fmt.Sprintf("%s/messages/send/interactive", apiURL)

	payload := InteractiveRequest{
		ChannelID: channelID,
		ContactID: contactID,
		Type:      "buttons",
		Body:      body,
		Buttons:   buttons,
	}

	respBody, err := makeRequest("POST", url, payload)
	if err != nil {
		return nil, err
	}

	var message Message
	if err := json.Unmarshal(respBody, &message); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &message, nil
}

// sendListMessage sends a list message with sections to a contact
func sendListMessage(channelID, contactID, body, buttonText string, sections []ListSection) (*Message, error) {
	url := fmt.Sprintf("%s/messages/send/interactive", apiURL)

	payload := InteractiveRequest{
		ChannelID:  channelID,
		ContactID:  contactID,
		Type:       "list",
		Body:       body,
		ButtonText: buttonText,
		Sections:   sections,
	}

	respBody, err := makeRequest("POST", url, payload)
	if err != nil {
		return nil, err
	}

	var message Message
	if err := json.Unmarshal(respBody, &message); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &message, nil
}

// sendQuickReplyMessage sends a quick reply message to a contact
func sendQuickReplyMessage(channelID, contactID, body string, buttons []Button) (*Message, error) {
	url := fmt.Sprintf("%s/messages/send/interactive", apiURL)

	payload := InteractiveRequest{
		ChannelID: channelID,
		ContactID: contactID,
		Type:      "quick_reply",
		Body:      body,
		Buttons:   buttons,
	}

	respBody, err := makeRequest("POST", url, payload)
	if err != nil {
		return nil, err
	}

	var message Message
	if err := json.Unmarshal(respBody, &message); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &message, nil
}

func main() {
	// Validate configuration
	if apiToken == "" {
		fmt.Println("Error: SENDSEVEN_API_TOKEN environment variable is required")
		os.Exit(1)
	}

	if tenantID == "" {
		fmt.Println("Error: SENDSEVEN_TENANT_ID environment variable is required")
		os.Exit(1)
	}

	if channelID == "" {
		fmt.Println("Error: CHANNEL_ID environment variable is required")
		os.Exit(1)
	}

	if contactID == "" {
		fmt.Println("Error: CONTACT_ID environment variable is required")
		os.Exit(1)
	}

	// 1. Check channel capabilities first
	fmt.Printf("Checking capabilities for channel: %s\n", channelID)
	capabilities, err := checkChannelCapabilities(channelID)
	if err != nil {
		fmt.Printf("Warning: Could not check capabilities: %v\n", err)
		fmt.Println("Proceeding anyway...")
		fmt.Println()
	} else {
		fmt.Printf("Channel type: %s\n", capabilities.ChannelType)
		caps := capabilities.Capabilities
		fmt.Printf("  Buttons: %v\n", caps["interactive_buttons"])
		fmt.Printf("  Lists: %v\n", caps["interactive_lists"])
		fmt.Printf("  Quick Replies: %v\n", caps["quick_replies"])
		fmt.Println()
	}

	// 2. Send a button message
	fmt.Println("Sending button message...")
	buttons := []Button{
		{ID: "yes", Title: "Yes"},
		{ID: "no", Title: "No"},
		{ID: "maybe", Title: "Maybe Later"},
	}

	message, err := sendButtonMessage(
		channelID,
		contactID,
		"Would you like to proceed with your order?",
		buttons,
	)
	if err != nil {
		fmt.Printf("Button message failed: %v\n\n", err)
	} else {
		fmt.Println("Button message sent successfully!")
		fmt.Printf("  ID: %s\n", message.ID)
		fmt.Printf("  Status: %s\n\n", message.Status)
	}

	// 3. Send a list message
	fmt.Println("Sending list message...")
	sections := []ListSection{
		{
			Title: "Electronics",
			Rows: []ListRow{
				{ID: "phones", Title: "Phones", Description: "Latest smartphones"},
				{ID: "laptops", Title: "Laptops", Description: "Portable computers"},
			},
		},
		{
			Title: "Accessories",
			Rows: []ListRow{
				{ID: "cases", Title: "Cases", Description: "Protective cases"},
				{ID: "chargers", Title: "Chargers", Description: "Fast chargers"},
			},
		},
	}

	message, err = sendListMessage(
		channelID,
		contactID,
		"Browse our product catalog:",
		"View Products",
		sections,
	)
	if err != nil {
		fmt.Printf("List message failed: %v\n\n", err)
	} else {
		fmt.Println("List message sent successfully!")
		fmt.Printf("  ID: %s\n", message.ID)
		fmt.Printf("  Status: %s\n\n", message.Status)
	}

	// 4. Send a quick reply message
	fmt.Println("Sending quick reply message...")
	quickReplies := []Button{
		{ID: "excellent", Title: "Excellent"},
		{ID: "good", Title: "Good"},
		{ID: "poor", Title: "Poor"},
	}

	message, err = sendQuickReplyMessage(
		channelID,
		contactID,
		"How would you rate our service today?",
		quickReplies,
	)
	if err != nil {
		fmt.Printf("Quick reply message failed: %v\n", err)
	} else {
		fmt.Println("Quick reply message sent successfully!")
		fmt.Printf("  ID: %s\n", message.ID)
		fmt.Printf("  Status: %s\n", message.Status)
	}
}
